drivers/ipm/ipm_ivshmem.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2023 Linaro.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT linaro_ivshmem_ipm
+
+#include <stdint.h>
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/ipm.h>
+#include <zephyr/drivers/virtualization/ivshmem.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ipm_ivshmem, CONFIG_IPM_LOG_LEVEL);
+
+struct ivshmem_ipm_data {
+	ipm_callback_t cb;
+	void *user_data;
+};
+
+struct ivshmem_ipm_config {
+	const struct device *ivshmem_dev;
+};
+
+static void ivshmem_ipm_event_loop_thread(void *arg, void *p2, void *p3)
+{
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	unsigned int poll_signaled;
+	int ivshmem_vector_rx;
+	struct k_poll_signal sig;
+	struct k_poll_event events[] = {
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,
+					 K_POLL_MODE_NOTIFY_ONLY,
+					 &sig),
+	};
+
+	const struct device *dev = (const struct device *)arg;
+	struct ivshmem_ipm_data *dev_data = (struct ivshmem_ipm_data *)dev->data;
+	struct ivshmem_ipm_config *dev_cfg = (struct ivshmem_ipm_config *)dev->config;
+
+	k_poll_signal_init(&sig);
+	int ret = ivshmem_register_handler(dev_cfg->ivshmem_dev, &sig, 0);
+
+	if (ret < 0) {
+		LOG_ERR("registering handlers must be supported: %d\n", ret);
+		k_panic();
```

Comment:
do you really need to panic here?

Reply:
From what I know of the IVSHMEM driver, failure from registering handlers is a unrecoverable error.